README.md
# digital-pligtaflevering-aviser-tools

[![wercker status](https://app.wercker.com/status/8696b835217d47b14f71175a72ad818c/s/ "wercker status")](https://app.wercker.com/project/byKey/8696b835217d47b14f71175a72ad818c)

Quicklinks:

* <a href="http://localhost:7880/fedora/objects/?pid=true&title=true&identifier=true&terms=DPA+Roundtrip&query=&maxResults=80">80 first DPA roundtrips</a>
* <a href="http://localhost:7880/fedora/objects?pid=true&title=true&identifier=true&terms=&query=&maxResults=80">80 first results in vagrant fedora</a>
* <a href="http://localhost:7880/fedora/objects?pid=true&title=true&identifier=true&terms=*pdf&query=&maxResults=80">80 first PDF files</a>
* <a href="http://localhost:58709/var/reference1pillar/dpaviser/folderDir/">Root of bitrepository quickstart pillar 1</a>
* <a href="http://localhost:58608/newspapr/sbsolr/#/collection1/query">SBOI summa</a>.

___

Digital Pligtaflevering af Aviser (DPA), autonomous preservation tools.

This document give an overview of the project and history.

Please also see:

* <a href="DEVELOPERS-GUIDE.md">Developers guide to DPA</a>
* ...


## Background

All Danish newspapers published are collected in physical form in two
copies and preserved by Statens Avissamling
(https://www.statsbiblioteket.dk/nationalbibliotek/adgang-til-samlingerne/aviser/StatensAvissamling).
The traditional procedure has then been microfilming the newspapers
and make the microfilms available to the general public.  A separate
project - https://blog.avisdigitalisering.dk/ - has digitized the old
microfilms and made them available through
http://www2.statsbiblioteket.dk/mediestream/avis We now want to start
collecting new newspapers as PDF's instead.  This is what DPA is
about, and is based on the experiences gained.

Official documentation is at
https://www.statsbiblioteket.dk/om-statsbiblioteket/udbud/levering-af-digitale-aviser

## Minimal Effort Ingest

The traditional mindset for digital preservation is to do all the work
for a given item, collecting metadata, normalizing files, ensuring
consistency and so on, and THEN store the success for eternity.  This
process is called "_ingestion_".  We have found that the amount of
work necessary to properly handle a given collection from end to end
is typically rather large, regardless of the size of the collection,
which mean that especially small collections are delayed much longer
than originally expected.

Our conclusion has been that it is more important to get the
collections preserved as they are and then doing the rest when we have
the time, than getting everything done and tying a bow on the virtual
box before being put on the digital shelf.

Official project page is at http://en.statsbiblioteket.dk/about-the-library/projects-1/minimal-effort-ingest

Steps relevant for PDF's could be:

* _Ingest_ PDF's (preserve them in Bitmagasinet)
* Analyse if the PDF can be rendered to a series of images, one per page.
* Check if the PDF is suitable for digital preservation.  An example could be if all fonts and images are present
inside the PDF
or if any need to be retrieved from the internet.  Such a resource may go away at any time, making the PDF incomplete.
If we receive such PDF's a human must take action.
* Create reports on deliveries to get an overview of how the process is going.

The important part is that after the initial ingest the steps can be
taken whenever the necessary resources are available, or re-taken if
found necessary.  New steps can also be added and run as appropriate
at any time later.

This only modifies the OASIS model of digital preservation slightly.

## Design

!(doc/images/IMG_9068.jpg "System outline 2016-10-24")


At SB (Statsbiblioteket - _State and University Library_) at the time
of design of DPA we work with two backend legacy systems:

1. "DOMS" (_Digital Object Management System_) - our homegrown
metadata repository based on Fedora Commons 3 (not the operating
system) for metadata storage in RDF form and Summa for searching.

2. "Bitmagasinet" (_Bit Repository_) - our long term bit preservation
system in collaboration with several other Danish institutions which
is where the actual files are stored.
https://sbforge.org/display/BITMAG/The+Bit+Repository+project

All information _about_ the files is stored in DOMS.  All the _actual_
files are stored in Bitmagasinet, and - based on a discussion TRA had
with KTC and ABR on 2016-07-14 - we need to utilize the SB-specific
implementation of the Bitmagasinet-pillar to access the physical files
efficiently similar to what has been done in Avisprojektet.

The steps taken so far on a given PDF is stored as metadata in the
form of PREMIS events.

Each step taken is implemented as a traditional stand-alone Java
application which regularily asks DOMS through Summa if there is any
PDF's ready to be worked on, and if any, work on those.  The success of
the work is added as an event and/or a datastream if appropriate for
later.

The workflow a PDF must pass through is defined by configuring the
steps so if step B must run after step A, then step B looks for the
PREMIS event that A has been run successfully.  This implicit assembly
line way of thinking allows us to avoid a full blown centralized
workflow engine and keep the complexity lower.

## Invocation

As of 2016-07-14 the only runnable code is the Main test in dpa-doms.
TRA reached the milestone of actually storing VeraPDF output in the
event and in the "VERAPDF" datastream today.  Note that both the doms-installer
vagrant machine and the VeraPDF rest server must be running for the
Main test to complete.

Plan from here is slowly migrating the Main test to the dpa-harness
module while building a Dependency Injection aware harness utilizing
Dagger 2 injecting all the needed constant strings as well as joining
the various components correctly.  When that work is done we should
be able to flesh out much better how the deliveries will look
and switch to work on one of those.


# Vagrant

In order to do development against a developer-local instance of
DOMS and Bitrepository an Ubuntu 14.04-based Vagrant image is used. 
Vagrant was chosen in order to be able to do rollbacks to snapshots
in VirtualBox.
[See vagrant/README.md for further information](vagrant/README.md).
 



# Flow

Summarized discussion with KFC 2016-07-01:

_The overall "keep running autonomous components on suitable items"
approach is only valid as long as no problems has been found.  If a
problem is found, the item is marked with a special event indicating
"manual triage necessary" which prohibits further automatic processing
until a human has investigated.  What happens then will be fleshed out
later._

After ingest, we run VeraPDF with the least restrictive profile on
each PDF.  The output will be similar to:

    {"totalAssertions":250,"pdfaflavour":"PDFA_1_B","testAssertions":[],"compliant":true}

For failures the `testAssertions` list will contain details.  This
output will be stored as metadata for the PDF-file, and an "veraPDF
has been run" event added.

Another autonomous component is responsible for analyzing the VeraPDF
output (in order to decouple the invocation - which might be expensive
- from the analysis).  The "Digital Bevaringsgruppe" has a list of
which issues to consider important.

If the PDF does not have any of these, it is given the "verapdf approved" event
and we are done.

If it does, then the "manual triage necessary" event is added, and the PDF must
be triaged by a human before automatic processing can continue.

# Manual triaging

If a "manual triage necessary" event is present on a PDF item in DOMS,
a human need to decide what needs to be done for the item.  (Tooling
for this needs to be discussed)

An autonomous component may not work on the item until this has been done.

The human may find one of (currently) three things:

1. The PDF is considered acceptable as is.  The "verapdf approved"
event is added manually, overriding "manual triage necessary" allowing
automatic processing to continue.

2. The PDF is found to require code alterations in order to be
processed correctly.  This requires involving developers and releasing
a new version.

3. The PDF is found to be unusable.  The human initiates actions that
procures a new and more usable PDF to replace the current file.  As
this is expected to be a very rare situation, the exact steps to take
are currently undefined.


# DOMS model created by Ingester based on Infomedia delivery format

The sample delivery "dl_20160811_rt1" (in delivery-samples/ folder) contains a single
test newspaper named "verapdf" with 9 physical pages and 1 article ready to be ingested.
The transfer_acknowledged and transfer_complete files are empty and ignored by the ingester.

    dl_20160811_rt1/
    dl_20160811_rt1/md5sums.txt
    dl_20160811_rt1/transfer_acknowledged
    dl_20160811_rt1/transfer_complete
    dl_20160811_rt1/verapdf
    dl_20160811_rt1/verapdf/articles
    dl_20160811_rt1/verapdf/articles/20160811_verapdf_article_5c649fe5.xml
    dl_20160811_rt1/verapdf/pages
    dl_20160811_rt1/verapdf/pages/20160811_verapdf_page001.pdf
    dl_20160811_rt1/verapdf/pages/20160811_verapdf_page001.xml
    dl_20160811_rt1/verapdf/pages/20160811_verapdf_page002.pdf
    dl_20160811_rt1/verapdf/pages/20160811_verapdf_page002.xml
    dl_20160811_rt1/verapdf/pages/20160811_verapdf_page003.pdf
    dl_20160811_rt1/verapdf/pages/20160811_verapdf_page003.xml
    dl_20160811_rt1/verapdf/pages/20160811_verapdf_page004.pdf
    dl_20160811_rt1/verapdf/pages/20160811_verapdf_page004.xml
    dl_20160811_rt1/verapdf/pages/20160811_verapdf_page005.pdf
    dl_20160811_rt1/verapdf/pages/20160811_verapdf_page005.xml
    dl_20160811_rt1/verapdf/pages/20160811_verapdf_page006.pdf
    dl_20160811_rt1/verapdf/pages/20160811_verapdf_page006.xml
    dl_20160811_rt1/verapdf/pages/20160811_verapdf_page007.pdf
    dl_20160811_rt1/verapdf/pages/20160811_verapdf_page007.xml

The DOMS model constructed looks like:

    +---------------------+
    ! "dl_20160811" + "1" !
    +-+-------------------+
      !
      +>+-----------+
        | "verapdf" |
        +-+---------+
            |
            +>+------------+
            ! ! "articles" !
            ! +-+----------+
            !   !
            !   +>+-------------------------------------------+
            !     ! "20160811_verapdf_article_5c649fe5"       !
            !     ! Datastream "XML": bytes in .XML file      ! 
            !     +-------------------------------------------+
            !
            +>+------------+
              ! "pages"    !
              +-+----------+
                !
                +>+-------------------------------------------+
                ! ! "20160811_verapdf_page001"                !
                ! ! Datastream "XML": bytes in .XML file      !
                ! +-+-----------------------------------------+
                !   ! NOTE: hasPart relation
                !   +>+-----------------------------------------------+
                !     ! "20160811_verapdf_page001.pdf"                ! 
                !     ! Physical file stored in the Bitrepository.    !
                !     ! Datastream "CONTENTS" points to bitfinder URL !
                !     +-----------------------------------------------+
                !
                +>+------------------------------------....
                ! ! "20160811_verapdf_page002".....
                ! ! Datastream "XML....
                ! +--- ...
                !

FIXME:  Move description from Javadoc into here. 

All arrows except one indicate a "hasPart" relation on the parent node
(typically directory) from the parent node to the child node.  There
is not a relation on the child pointing to its parent.  The exception
is the relation from the "page" node to the PDF file (which is
stored externally in the Bitrepository) which is a "hasFile" relation.


This is the expected Event-flow of the newspaper-deliveries:

Event UI is a userinterface that can be used to modify the Event-flow for a specific delivery

    +-----------------------------+
    ! Create Delivery             !  (Data_Received)
    +-----------------------------+
                  !                                                     +----------+
                 < >---Fail---------------------------------------------! Event UI !
                  !                                                     +----------+
                  V
    +-----------------------------+
    ! Ingest                      !  (Data_Archived)
    +-----------------------------+
                  !                                                     +----------+
                 < >---Fail---------------------------------------------! Event UI !
                  !                                                     +----------+
                  V
    +-----------------------------+
    ! Generate checksum           !  (Checksumfile_regenerated)
    +-----------------------------+
                  !                                                     +----------+
                 < >---Fail---------------------------------------------! Event UI !
                  !                                                     +----------+
                  V
    +-----------------------------+
    ! Check generated checksum    !  (Checksumfile_checked)
    +-----------------------------+
                  !                                                     +----------+
                 < >---Fail---------------------------------------------! Event UI !
                  !                                                     +----------+
                  V
    +-----------------------------+
    ! Generate statistics         !  (Statistics_generated)
    +-----------------------------+
                  !                                                     +----------+
                 < >---Fail---------------------------------------------! Event UI !
                  !                                                     +----------+
                  V
    +-----------------------------+
    ! Create VeraPDF              !  (VeraPDF_Invoked)
    +-----------------------------+
                  !                                                     +----------+
                 < >---Fail---------------------------------------------! Event UI !
                  !                                                     +----------+
                  V
    +-----------------------------+
    ! Check VeraPDF               !  (VeraPDF_Analyzed)
    +-----------------------------+
                  !                                                     +----------+
                 < >---Fail---------------------------------------------! Event UI !
                  !                                                     +----------+
                  V
    +-----------------------------+
    ! Generate Del. pattern match !  (Newspaper_Weekdays_Analyzed)
    +-----------------------------+
                  !                                                     +----------+
                 < >---Fail---------------------------------------------! Event UI !
                  !                                                     +----------+
                  !        
                  V        
    +-----------------------------+
    ! Manual control UI           !
    +-----------------------------+                                    



# VeraPDF

VeraPDF is a PDF/A validator suite which we use to get reports of the
preservation quality of the PDFs ingested.

Due to various problems it needs to run in a separate JVM.  See <https://sbprojects.statsbiblioteket.dk/jira/browse/DPA-124>
for details.  It was found that the easiest way to get a production ready 
REST service was to improve on <https://github.com/veraPDF/veraPDF-rest>.

Currently it resides in 

    git clone https://github.com/statsbiblioteket/veraPDF-rest 
    
until the changes are merged back upstream.  Note that as of 2018-07-09 this is on the
TRA branch which github is configured to know!

Ensure the artifact is available to Maven (e.g. installed in the ~/.m2 local repository) 
before building the vagrant box.
 

To get the VeraPDF-rest shaded jar for local use or to deploy to stage/prod use: 

    mvn -B -q dependency:copy -Dartifact=org.verapdf:verapdf-rest:0.2.0-SNAPSHOT:jar:shaded -DoutputDirectory=. -Dmdep.stripClassifier -Dmdep.stripVersion 

to download to verapdf-rest.jar to the current directory.  

When deployed use an invocation similar to:

    java -Ddw.server.applicationConnectors[0].port=8090 -Ddw.server.adminConnectors[0].port=8091 -jar verapdf-rest.jar server > logfile

in a restarting wrapper.  Use Java 8!




(2017-03-20:  This information is outdated.  DPA-24 include a request for this document to be updated)

As of 2016-06-30 the initial prototype uses the VeraPDF REST
interface.

With the current state of the VeraPDF sources the easiest way to get
an executable jar is:

Go to the latest build of the rest service on the Open Preservation jenkins:

[http://jenkins.openpreservation.org/job/veraPDF-rest/lastBuild/org.verapdf$verapdf-rest/](http://jenkins.openpreservation.org/job/veraPDF-rest/lastBuild/org.verapdf$verapdf-rest/)

Download the jar file, e.g. verapdf-rest-0.1.22.jar.

Run it manually with

    java -jar verapdf-rest-0.1.22.jar server

By default it listens on port 8000.

# DOMS

For development a full "server vagrant" with DOMS, Zookeeper etc has been
adapted from the doms-installer project and is placed in doms/doms-installer.
This is not integrated in the main pom.xml as this requires insane bandwidth
and read access to the internal SB Nexus instance.

NOTE:  THIS IS NOT PRODUCTION QUALITY.  JUST ENOUGH TO ALLOW LOCAL WORK.

See doms/doms-installer/README.md for full instructions.  The following are just the commands
needed after initial configuration to get up and running.

Clean and download necessary packages from internal build server:

    mvn -f doms/doms-installer/pom.xml clean package

Get vagrant services up and running:

    cd doms/doms-installer/src/vagrant
    vagrant up
    vagrant ssh -c "nohup bash -x /vagrant/install_doms.sh; nohup bash -x /vagrant/setup-newspapers.sh"

DOMS with friends are now running.  Do NOT stop this virtual machine
as the services do not come up propely.  Pause it and restart it as
needed.

Access Fedora (fedoraAdmin/fedoraAdminPass):

[http://localhost:7880/fedora/objects](http://localhost:7880/fedora/objects)

The default DOMS content created contains two newspaper batches that
is triggered by the search criteria, and is simulated to be run on
dummy PDF files stored as test resources.  VeraPDF is invoked as a
REST service, and the output stored in the event and as a datastream
named VERAPDF on the newspaper roundtrip object.


# Running visualvm on DOMS

ssh in with X11-forwarding active

    vagrant ssh -- -Y
    
Install visualVM and run it.

    sudo apt install visualvm

    jvisualvm

# Creating *.verapdf files on 

Due to the slowness of VeraPDF 0.26 TRA has augmented the FilesystemDeliveryIngester to add `.verapdf` files as 
datastreams in the same way that `.xml` files are.  They can be precomputed with this command
on a local copy of the files as `scape@miaplacidus`.

    find . -name '*.pdf' -print0 | sort | $HOME/tra/gnu/bin/parallel -0 --bar --gnu -j35 ' $HOME/tra/verapdf/verapdf -f 1b --format mrr {} > $(dirname {})/$(basename {} .pdf).verapdf'

(GNU parallels is installed in $HOME/tra/gnu/bin, and a installed copy of $HOME/verapdf is copied to $HOME/tra)




Changelog.md
tools/dpa-tools-deployment/for-deployment/Changelog.md

main/dpa-doms/Main.java
package doms;

import dk.statsbiblioteket.digital_pligtaflevering_aviser.bitrepository.LookupFileInfoForID;
import dk.statsbiblioteket.digital_pligtaflevering_aviser.doms.DomsEvent;
import dk.statsbiblioteket.digital_pligtaflevering_aviser.doms.DomsItem;
import dk.statsbiblioteket.digital_pligtaflevering_aviser.doms.DomsQuery;
import dk.statsbiblioteket.digital_pligtaflevering_aviser.doms.QuerySpecification;
import dk.statsbiblioteket.digital_pligtaflevering_aviser.model.Task;
import dk.statsbiblioteket.digital_pligtaflevering_aviser.verapdf.VeraPDFWebServices;
import dk.statsbiblioteket.medieplatform.autonomous.DomsEventStorage;
import dk.statsbiblioteket.medieplatform.autonomous.DomsEventStorageFactory;
import dk.statsbiblioteket.medieplatform.autonomous.Item;
import dk.statsbiblioteket.medieplatform.autonomous.ItemFactory;
import dk.statsbiblioteket.medieplatform.autonomous.PremisManipulatorFactory;
import dk.statsbiblioteket.medieplatform.autonomous.SBOIEventIndex;
import org.junit.Test;

import javax.inject.Provider;
import java.io.IOException;
import java.io.InputStream;
import java.util.List;
import java.util.function.Function;
import java.util.stream.Collectors;
import java.util.stream.Stream;

import static java.util.Arrays.asList;
import static java.util.Collections.emptyList;

/**
 *
 */
public class Main {
    public static void main(String[] args) throws Exception {
    }

    @Test
    public void Main() throws Exception {
        //org.bitrepository.common.filestore.FileInfo fileInfo = new org.bitrepository.common.filestore.FileInfo();

        ItemFactory<Item> itemFactory = id -> new Item(id);

        final Provider<DomsEventStorage<Item>> domsEventStorageProvider = getDomsEventStorageProvider(
                itemFactory,
                "http://localhost:7880/fedora",
                "http://localhost:7880/pidgenerator-service");

        final DomsEventStorage<Item> domsEventStorage = domsEventStorageProvider.get();

        SBOIEventIndex<Item> index = new SBOIEventIndex<>(
                "http://localhost:58608/newspapr/sbsolr/",
                new PremisManipulatorFactory<>(PremisManipulatorFactory.TYPE, itemFactory),
                domsEventStorage,
                10
        );

        // get inputstream for an id in e.g. the bitrepository.
        Function<String, InputStream> inputStreamFor = id -> {
            try {
                return new LookupFileInfoForID().apply(id).getInputstream();
            } catch (IOException e) {
                throw new RuntimeException("could not get inputstream for " + id, e);
            }
        };

        QuerySpecification specification = new QuerySpecification(
                asList("Data_Received"),
                emptyList(),
                emptyList(),
                emptyList(),
                false
        );

        DomsQuery<Item> q = new DomsQuery<>(domsEventStorage, index);

        Stream<DomsItem> itemStream = q.query(specification);

        Task<DomsItem, String> task = item -> {
            String bitrepositoryId = "";  // FIXME: bitrepository ingester must work first.
            InputStream inputStream = inputStreamFor.apply(bitrepositoryId);

            VeraPDFWebServices veraPDFWebServices = new VeraPDFWebServices("http://localhost:8080/api");
            String fileName = "fake original filename.pdf"; // item.getOriginalItem().getDomsID();
            String outputFromVeraPDf = veraPDFWebServices.validate(inputStream, fileName, "1b");
            item.datastreams().put("VERAPDF", outputFromVeraPDf);

            DomsEvent event = new DomsEvent("dummy agent name for now", outputFromVeraPDf, item.toString(), false);
            item.events().add(event);

            return outputFromVeraPDf;
        };

        List<String> result = itemStream.map(task).collect(Collectors.toList());

        System.out.println(result); // FIXME
    }

    public Provider<DomsEventStorage<Item>> getDomsEventStorageProvider(ItemFactory<Item> itemFactory, String
            fedoraLocation, String pidGeneratorLocation) {

        DomsEventStorageFactory<Item> domsEventStorageFactory = new DomsEventStorageFactory<>();
        domsEventStorageFactory.setFedoraLocation(fedoraLocation);
        domsEventStorageFactory.setPidGeneratorLocation(pidGeneratorLocation);
        //domsEventStorageFactory.setUsername(properties.getProperty(DOMS_USERNAME));
        //domsEventStorageFactory.setPassword(properties.getProperty(DOMS_PASSWORD));
        domsEventStorageFactory.setItemFactory(itemFactory);

        return () -> {
            try {
                return domsEventStorageFactory.createDomsEventStorage();
            } catch (Exception e) {
                throw new RuntimeException("get() failed", e);
            }
        };
    }
}
